### PR TITLE
Refactoring: select tests

### PR DESCRIPTION
**Pull request by toaster**
### Description:

This PR makes the select tests independent from the implementation details.
This makes the menu refactoring a lot easier.

Also included:
* fix a refresh bug in the shadow implementation
* testApp applies theme to overlays
* move testTheme into test package
* overlay support for software painter
* `test.WaitForThemeToBeApplied()`

### Checklist:

- [x] Tests included.
- [x] Lint and formatter run with no errors.
- [x] Tests all pass.

#### Where applicable:

- [x] Public APIs match existing style.

### CODE REVIEW

File: test/testapp.go [hunk 1]
```diff
@@ -50,29 +52,31 @@ func (a *testApp) UniqueID() string {
...
 		return
 	}
 	content.Refresh()
 
-	if wid, ok := content.(fyne.Widget); ok {
-		for _, o := range wid.CreateRenderer().Objects() {
-			a.applyThemeTo(o, canv)
+	switch x := content.(type) {
+	case fyne.Widget:
+		for _, o := range cache.Renderer(x).Objects() {
```
Comment (andydotxyz) [inline test/testapp.go:63]: Wouldn't `test.WidgetRenderer()` be better, not adding dependency on cache package.

File: test/util.go [hunk 2]
```diff
@@ -96,12 +97,34 @@ func TypeOnCanvas(c fyne.Canvas, chars string) {
 	typeChars([]rune(chars), c.OnTypedRune())
 }
 
+// WaitForThemeToBeApplied waits for the current theme to be applied to the current app.
+func WaitForThemeToBeApplied(t *testing.T) {
```
Comment (andydotxyz) [inline test/util.go:101]: How about making this `ApplyTheme`, take the theme parameter and do the settings apply as well.
Every call of it follows a setTheme, so why not bring it into this method :)

File: test/util.go [hunk 3]
```diff
@@ -96,12 +97,34 @@ func TypeOnCanvas(c fyne.Canvas, chars string) {
...
+}
+
 // WidgetRenderer allows test scripts to gain access to the current renderer for a widget.
 // This can be used for verifying correctness of rendered components for a widget in unit tests.
 func WidgetRenderer(wid fyne.Widget) fyne.WidgetRenderer {
 	return cache.Renderer(wid)
 }
 
+// WithTestTheme runs a function with the testTheme temporarily set.
+func WithTestTheme(t *testing.T, f func()) {
```
Comment (andydotxyz) [inline test/util.go:116]: Open question here (and I know I named it this way, but) with the prefix now would `test.WithTheme(...)` be clear enough?

File: test/testapp.go (hunk 1 above)
Comment (toaster) [reply test/testapp.go:63]: Well, it would not really reduce the dependency of the package but okay.

File: test/util.go (hunk 3 above)
Comment (toaster) [reply test/util.go:116]: I don’t think so. When reading `test.WithTheme(f)` I read that this is the test helper `WithTheme` and it expects at least a `Theme` to use.

File: test/testapp.go (hunk 1 above)
Comment (andydotxyz) [reply test/testapp.go:63]: I meant that this has a new dependency on cache, that line of code would go away - hiding the internal details.

File: test/util.go (hunk 3 above)
Comment (andydotxyz) [reply test/util.go:116]: Ah I see your point. I had imagined that but missed the parameter list sorry :)

**Pull request merged**